### PR TITLE
Containerize services & orchestrate with docker-compose

### PR DESCRIPTION
**Pull request by cmyui**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+version: "3"
+
+services:
+  mysql:
+    image: mysql
+    environment:
+      MYSQL_USER: ${MYSQL_USER}
+      MYSQL_PASSWORD: ${MYSQL_PASSWORD}
+      MYSQL_DATABASE: ${MYSQL_DATABASE}
```
Comment (cmyui) [inline docker-compose.yml:9]: didn't know u could use env variables here - this is really cool

File: .env.example [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
-REDIS_DSN=redis://localhost
+HOST_PORT=443
 
-OSU_API_KEY=
+# XXX: change these values!
+MYSQL_USER=
+MYSQL_PASSWORD=
+MYSQL_DATABASE=
+# then copy the above values (user, password, database) to this:
+DB_DSN=mysql://user:password@mysql/database
```
Comment (cmyui) [inline .env.example:13]: i think i'd rather construct the dsn in the code rather than include it here as well - i'm fine storing user/pass/dbname in env

should do the same for redis too though

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+WORKDIR /
+COPY requirements.txt ./
+RUN pip install -r requirements.txt
+
+# copy the source code in last, so that it doesn't
+# repeat the previous steps for each change
+COPY . .
+
+ENTRYPOINT ["python3"]
+CMD ["main.py"]
```
Comment (cmyui) [inline Dockerfile:21]: we should run scripts/start.sh here rather than main.py directly

we should use https://github.com/vishnubob/wait-for-it for connected service healthchecks so we start everything in order

File: app/utils.py [hunk 4]
```diff
@@ -382,9 +382,9 @@ def ensure_local_services_are_running() -> int:
                 log("Unable to connect to mysql server.", Ansi.LRED)
                 return 1
 
-    if not os.path.exists("/var/run/redis/redis-server.pid"):
-        log("Unable to connect to redis server.", Ansi.LRED)
-        return 1
+    # if not os.path.exists("/var/run/redis/redis-server.pid"):
+    #     log("Unable to connect to redis server.", Ansi.LRED)
+    #     return 1
```
Comment (cmyui) [inline app/utils.py:387]: this stuff can all be removed once we have the healthchecks in the start.sh scriptr

File: .env.example (hunk 2 above)
Comment (tsunyoku) [reply .env.example:13]: +1 on that

File: Dockerfile (hunk 3 above)
Comment (tsunyoku) [reply Dockerfile:21]: +1 on that

File: app/utils.py (hunk 4 above)
Comment (tsunyoku) [reply app/utils.py:407]: +1 on that

File: docker-compose.yml (hunk 1 above)
Comment (kingdom5500) [reply docker-compose.yml:9]: yeah, and the mysql image from docker hub automatically generates the user and database and privileges and so on, which makes things nice and easy for the db init :D

File: .env.example (hunk 2 above)
Comment (kingdom5500) [reply .env.example:13]: completely agree

File: .env.example [hunk 5]
```diff
@@ -1,17 +1,25 @@
-# bancho.py supports both INET and UNIX sockets
-# INET: provide an IP for SERVER_ADDR, and a port for SERVER_PORT
-# UNIX: provide a filesystem path for SERVER_ADDR, and no SERVER_PORT
-SERVER_ADDR=/tmp/bancho.sock
-SERVER_PORT=
+# for nginx reverse proxy to work through the containers, the bancho
+# server must expose itself on port 80 to be accessed on http://bancho.
+SERVER_ADDR=0.0.0.0
```
Comment (tsunyoku) [inline .env.example:3]: unsure if i would go about putting default addr & port to docker values as not everyone will be using it. i’d atleast keep the comments explaining how to use unix/inet there too

File: .env.example [hunk 6]
```diff
@@ -42,6 +50,10 @@ DISCORD_AUDIT_LOG_WEBHOOK=
 # for debugging & development purposes.
 AUTOMATICALLY_REPORT_PROBLEMS=False
 
+# XXX: change these to their location on your host server:
+SSL_CERT_PATH=/etc/ssl/certs/fullchain.pem
```
Comment (tsunyoku) [inline .env.example:54]: this as a config option is kinda weird since most people use cf for ssl and so this isnt required

File: .gitignore [hunk 7]
```diff
@@ -6,6 +6,7 @@ _testing/
 *.lprof
 ext/GeoLite2-City.mmdb
 .env
+/ext/nginx.conf
```
Comment (tsunyoku) [inline .gitignore:9]: why /ext/ and not ext/

File: README.md [hunk 8]
```diff
@@ -45,97 +46,28 @@ we aim to minimize our dependencies, but still rely on ones such as
...
-# install bancho.py's python-specific dependencies
-python3.9 -m pip install -r requirements.txt
-```
-
-## creating a database for bancho.py
-you will need to create a database for bancho.py to store persistent data.
-
-the server uses this database to store metadata & logs, such as user accounts
-and stats, beatmaps and beatmapsets, chat channels, tourney mappools and more.
+these dependencies are all retrieved by and contained within docker containers.
```
Comment (tsunyoku) [inline README.md:49]: i wouldn’t straight up remove regular setup requirements; people should have the option for docker or manual setup

File: README.md [hunk 9]
```diff
@@ -45,97 +46,28 @@ we aim to minimize our dependencies, but still rely on ones such as
...
-```sh
-# you'll need to change:
-# - YOUR_DB_NAME
-# - YOUR_DB_USER
-
-# import bancho.py's mysql structure to our new db
-# this runs the contents of the file as sql commands.
-mysql -u YOUR_DB_USER -p YOUR_DB_NAME < migrations/base.sql
-```
+you also probably want to modify or remove the `web` service from the `docker-compose.yml`
```
Comment (tsunyoku) [inline README.md:69]: same here about docker and manual

File: README.md [hunk 10]
```diff
@@ -153,47 +85,43 @@ sudo certbot certonly \
     -d *.YOUR_DOMAIN
 ```
 
+put the key and certificate somewhere on your server, and then
```
Comment (tsunyoku) [inline README.md:88]: same again about ssl being mostly from cf and this not being necessary

File: README.md [hunk 11]
```diff
@@ -153,47 +85,43 @@ sudo certbot certonly \
...
 ```
 
 ## congratulations! you just setup an osu! private server
 
 if everything went well, you should be able to start your server up:
 
 ```sh
-# start the server
-./main.py
+# start the server. note: you may have to start the services
```
Comment (tsunyoku) [inline README.md:110]: again about docker and manual setup

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+      MYSQL_RANDOM_ROOT_PASSWORD: "true"
+    volumes:
+      - ./migrations/base.sql:/docker-entrypoint-initdb.d/init.sql:ro
+
+  redis:
+    image: redis
+
+  bancho:
```
Comment (tsunyoku) [inline docker-compose.yml:17]: wheres port exposure for bancho?

File: ext/nginx.conf.example [hunk 13]
```diff
@@ -13,11 +9,12 @@ server {
...
 	# XXX: you'll need to edit this to match your domain (mine's cmyui.xyz)
-	server_name ~^(?:c[e4]?|osu|a|b|api)\.cmyui\.xyz$;
+	server_name ~^(?:c[e4]?|osu|a|b|api)\.example\.com$;
 
-	# XXX: you'll need to edit this to match your certificate and key
-	ssl_certificate     /etc/letsencrypt/live/cmyui.xyz/fullchain.pem;
-	ssl_certificate_key /etc/letsencrypt/live/cmyui.xyz/privkey.pem;
+	# these paths are hardcoded in docker-compose.yml so now
+	# you don't need to change these unless you changed that.
+	ssl_certificate     /etc/ssl/certs/bancho.pem;
```
Comment (tsunyoku) [inline ext/nginx.conf.example:16]: once again certs should be not provided by default because of cf

File: ext/nginx.conf.example [hunk 14]
```diff
@@ -1,10 +1,6 @@
 # NOTE: if you wish to only connect using fallback, you can
 # remove all ssl related content from the configuration.
 
-upstream bancho {
```
Comment (tsunyoku) [inline ext/nginx.conf.example:4]: why is this removed when its still used further down?

File: ext/nginx.conf.example [hunk 15]
```diff
@@ -29,18 +26,31 @@ server {
...
 	# XXX: you'll need to edit this to match your domain (mine's cmyui.xyz)
-	server_name assets.cmyui.xyz;
+	server_name assets.example.com;
 
-	# XXX: you'll need to edit this to match your certificate and key
-	ssl_certificate     /etc/letsencrypt/live/cmyui.xyz/fullchain.pem;
-	ssl_certificate_key /etc/letsencrypt/live/cmyui.xyz/privkey.pem;
+	# these paths are hardcoded in docker-compose.yml so now
+	# you don't need to change these unless you changed that.
+	ssl_certificate     /etc/ssl/certs/bancho.pem;
```
Comment (tsunyoku) [inline ext/nginx.conf.example:52]: again about certs and cf

**Review by tsunyoku — CHANGES_REQUESTED** [review]

File: .env.example (hunk 5 above)
Comment (cmyui) [reply .env.example:3]: i think we can probably expect everyone to use this going forward - was planning on updating the readme as well

File: .env.example (hunk 6 above)
Comment (cmyui) [reply .env.example:54]: yeah i have an issue open for that https://github.com/osuAkatsuki/bancho.py/issues/264

for now this is ok

File: ext/nginx.conf.example (hunk 14 above)
Comment (cmyui) [reply ext/nginx.conf.example:4]: i think http;//bancho would be valid within the docker compose so this is no longer needed

File: .gitignore (hunk 7 above)
Comment (cmyui) [reply .gitignore:9]: (meaning my other ones should be updated)

**Comment by Sclafus** [comment]
Is this still worked on? Would be a nice addition to the project.

**Comment by cmyui** [comment]
> Is this still worked on? Would be a nice addition to the project.

hey, i've got a lot going on outside of this project at the moment (recently hired as a sr. software engineer at a startup, working towards moving into a new apartment, etc.) so i'm pretty out of time

because of that, i don't think i'll be able to get around to this any time soon; i'm definitely open to having people take over this pr/the project as a whole if anyone shows the potential to be able to manage it

it would indeed be a pretty cool addition. i'd also like to make the server stateless and store ephemeral data (e.g. player sessions) in redis, but that would take quite a bit of time i don't currently have

**Comment by tsunyoku** [comment]
maybe i can revive this

**Comment by cmyui** [comment]
being continued in #320
